IRremoteESP8266.cpp

```diff
@@ -1630,35 +1681,73 @@ bool ICACHE_FLASH_ATTR IRrecv::decodeRCMM(decode_results *results) {
   return true;
 }
 
-bool ICACHE_FLASH_ATTR IRrecv::decodePanasonic(decode_results *results) {
+// Decode the supplied Panasonic message.
+//
+// Args:
+//   results: Ptr to the data to decode and where to store the decode result.
+//   nbits:   Nr. of data bits to expect.
+//   strict:  Flag indicating if we should perform strict matching.
+// Returns:
+//   boolean: True if it can decode it, false if it can't.
+//
+// Note:
+//   Panasonic 48-bit protocol is a modified version of Kaseikyo.
+// Ref:
+//   http://www.remotecentral.com/cgi-bin/mboard/rc-pronto/thread.cgi?26152
+bool ICACHE_FLASH_ATTR IRrecv::decodePanasonic(decode_results *results,
+                                               uint16_t nbits,
+                                               bool strict) {
+  if (results->rawlen < 2 * nbits + 4)
+    return false;  // Not enough entries to be a Panasonic message.
+  if (strict && nbits != PANASONIC_BITS)
+    return false;  // Request is out of spec.
+
   unsigned long long data = 0;
-	int offset = 1;  // Dont skip first space
-  if (!matchMark(results->rawbuf[offset], PANASONIC_HDR_MARK)) {
+	int offset = 1;
+
+  // Header
+  if (!matchMark(results->rawbuf[offset++], PANASONIC_HDR_MARK))
     return false;
-  }
-  offset++;
-  if (!matchMark(results->rawbuf[offset], PANASONIC_HDR_SPACE)) {
+  if (!matchMark(results->rawbuf[offset++], PANASONIC_HDR_SPACE))
     return false;
-  }
-  offset++;
-  // decode address
-  for (int i = 0; i < PANASONIC_BITS; i++) {
-    if (!match(results->rawbuf[offset++], PANASONIC_BIT_MARK)) {
+
+  // Data
+  for (uint16_t i = 0; i < nbits; i++) {
+    if (!match(results->rawbuf[offset++], PANASONIC_BIT_MARK))
       return false;
-    }
-    if (match(results->rawbuf[offset],PANASONIC_ONE_SPACE)) {
-      data = (data << 1) | 1;
-    } else if (match(results->rawbuf[offset],PANASONIC_ZERO_SPACE)) {
-      data <<= 1;
-    } else {
+    if (match(results->rawbuf[offset],PANASONIC_ONE_SPACE))
+      data = (data << 1) | 1;  // 1
+    else if (match(results->rawbuf[offset],PANASONIC_ZERO_SPACE))
+      data <<= 1;  // 0
+    else
       return false;
-    }
     offset++;
```

Comment:
/me looks blankly at it for a minute.
D'oh! Of course it could.
Consider it done.